storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,6 +3579,238 @@ dberr_t fsp_sys_tablespace_validate()
+   usedInodes->insert_seg(
+     block->page.frame + PAGE_HEADER + PAGE_BTR_SEG_TOP);
+   usedInodes->insert_seg(
+     block->page.frame + PAGE_HEADER + PAGE_BTR_SEG_LEAF);
+ }
+ mtr.commit();
+ return err;
+}
+
+/** Free the extents, fragment page from the given inode
+@param iblock inode block
+@param inode  index node information
+@param space  system tablespace
+@param mtr    mini-transaction */
+dberr_t fseg_inode_free(
+  buf_block_t *iblock, fseg_inode_t *inode,
+  fil_space_t *space, mtr_t *mtr)
+{
+  dberr_t err= DB_SUCCESS;
+  xdes_t *descr= fseg_get_first_extent(inode, space, mtr, &err);
+  while (descr)
+  {
+    err= fseg_free_extent(inode, iblock, space,
+                          xdes_get_offset(descr), mtr);
+    if (err) return err;
+    descr= fseg_get_first_extent(inode, space, mtr, &err);
+    if (err) return err;
+  }
+
+  for (uint32_t i= 0; i < FSEG_FRAG_ARR_N_SLOTS; i++)
+  {
+    uint32_t page_no = fseg_get_nth_frag_page_no(inode, i);
+    if (page_no != FIL_NULL)
+    {
+      err= fseg_free_page_low(inode, iblock, space, page_no, mtr);
+      if (err) return err;
+    }
+  }
+  fsp_free_seg_inode(space, inode, iblock, mtr);
+  return err;
+}
+
+/** Free the unused segment
+@param boffset    offset of the list containing segment header
+@param space      system tablespace
+@param usedInodes Used system tablespace index node
+@param mtr        mini-transaction
+@return error code */
+dberr_t fsp_free_unused_seg(
+  ulint boffset, fil_space_t *space, inodeInfo *usedInodes,
+  mtr_t *mtr)
+{
+  dberr_t err= DB_SUCCESS;
+  buf_block_t *block= fsp_get_header(space, mtr, &err);
+  if (!block)
+    return err;
+  const uint32_t len= flst_get_len(block->page.frame + boffset);
+  fil_addr_t addr= flst_get_first(block->page.frame + boffset);
+  for (uint32_t i= len; i--; )
+  {
+    ut_ad(addr.boffset >= FIL_PAGE_DATA);
+    ut_ad(addr.boffset < block->physical_size() - FIL_PAGE_DATA_END);
+    block= buf_page_get_gen(
+      page_id_t(0, addr.page), 0, RW_SX_LATCH, nullptr, BUF_GET, mtr, &err);
+    if (!block)
+      return err;
+    for (uint32_t i= 0; i < FSP_SEG_INODES_PER_PAGE(srv_page_size); i++)
+    {
+      fseg_inode_t *inode= fsp_seg_inode_page_get_nth_inode(
+        block->page.frame, i);
+      uint16_t offset= uint16_t (inode - block->page.frame);
+      if (mach_read_from_8(FSEG_ID + inode) == 0)
+        continue;
+      if (!usedInodes->find(addr.page, offset))
+      {
+        err= fseg_inode_free(block, inode, space, mtr);
+        if (err)
+          return err;
+      }
+    }
+    addr= flst_get_next_addr(block->page.frame + addr.boffset);
+  }
+  ut_ad(addr.page == FIL_NULL);
+  return err;
+}
+
+/** Remove the unused segment in system tablespace
+@return error code */
+dberr_t fsp_garbage_collect()
+{
+  if (!trx_sys.is_undo_empty())
+  {
+    sql_print_warning("InnoDB: Can't free the unused segment "
+                      "in system tablespace due to undo log exists.");
+    return DB_ERROR;
+  }
```

Comment:
Could we rename this function to `fil_space_t::garbage_collect()`? Invoked as `fil_system.sys_space->garbage_collect()`.